### PR TITLE
Fix dockerfile parsing bug due to missing newline

### LINKED ISSUE #10084: limit memory error on a docker file | Parsing bug
**Describe the bug**
A dockerfile that ends with ends a LABEL instruction causes limit memory error , I know that no dockerfile supposed to end with a LABEL , but my automatic scan can handle those type of files , and i cant not just exclude them or add them to .getignore because i cant always know their name in advance , on semgrep forum they suggested that i open a bug!

**To Reproduce**
Steps to reproduce the behavior, (https://semgrep.dev/playground/r/L1UqpL7/mona_pers1996_personal_org.npm-install-without-production-flag)

**Expected behavior**
I expect that this behavior gets parsed to taken care of with out causing a limit on memory 

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [x] P1: important to fix or quite annoying
- [ ] P2: regular bug that should get fixed

### PR DESCRIPTION
**Pull request by mjambon**
This isn't a fix in tree-sitter-dockerfile as it should be but it's the same workaround we had in place for parsing patterns. Dockerfile patterns often lack a trailing newline, but regular dockerfiles may also lack a trailing newline. This PR adds the missing newline to the input if it's missing. Such errors where tree-sitter inserts a token were silently ignored in semgrep as well as in our parsing tests. I added a field `tolerated_errors` to the parsing result that allows our tests to be strict about this. We now have two tests that would fail without the newline fix. Semgrep still ignores "missing token" errors (saves falling back to a slower pfff parser, as already annotated in the code).

There is a deeper issue reported by https://github.com/semgrep/semgrep/issues/10084 where their input (our new test file `no-trailing-newline-segfault.dockerfile`) causes a segmentation fault. It is reported incorrectly as an out-of-memory error by the Semgrep Playground but this is not very important. Adding the new line at the end of the input allows for proper parsing without a segfault but this should be investigated further to determine if it's a bug in tree-sitter-dockerfile's `scanner.c`, in the OCaml/tree-sitter bindings, or something else.

Update: the issue is known and tracked at https://github.com/camdencheek/tree-sitter-dockerfile/issues/22. It should be fixed with a tree-sitter upgrade. (thx @kopecs)

Fixes https://github.com/semgrep/semgrep/issues/10084

semgrep-pro: https://github.com/semgrep/semgrep-proprietary/pull/1469

### CODE REVIEW

**Comment by kopecs** [comment]
Probably related: https://github.com/camdencheek/tree-sitter-dockerfile/issues/22

**Review by ajbt200128 — COMMENTED** [review]
JS related stuff looks fine. Someone else should approve this though since I haven't worked with the docker parser

File: languages/dockerfile/tree-sitter/Parse_dockerfile_tree_sitter.ml [hunk 1]
```diff
@@ -1032,9 +1032,23 @@ let source_file (env : env) (xs : CST.source_file) =
 (* Entry points *)
 (*****************************************************************************)
 
+(* tree-sitter-dockerfile parser requires a trailing newline.
+   Not sure if it's intentional but we add one to simplify user
+   experience.
+*)
+let ensure_trailing_newline str =
```
Comment (kopecs) [inline languages/dockerfile/tree-sitter/Parse_dockerfile_tree_sitter.ml]: Could you include a link to the upstream issue here for reference, especially if we think this is removable after a tree sitter version upgrade?

Comment (mjambon) [reply languages/dockerfile/tree-sitter/Parse_dockerfile_tree_sitter.ml]: Good idea. I added a comment with the relevant details.

**Review by DrewDennison — APPROVED** [review]

**Pull request merged**